tool/ci/pkg/environment/environment.go

```diff
@@ -0,0 +1,289 @@
+	unmarshalRevs unmarshalReviewersFn
+}
+
+// Environment contains information about the environment
+type Environment struct {
+	Client           *github.Client
+	PullRequest      *PullRequestMetadata
+	token            string
+	reviewers        map[string][]string
+	defaultReviewers []string
+	action           string
+}
+
+// PullRequestMetadata is the current pull request metadata
+type PullRequestMetadata struct {
+	Author    string
+	RepoName  string
+	RepoOwner string
+	Number    int
+	HeadSHA   string
+	BaseSHA   string
+	// Only used for pull request review events
+	Reviewer   string
+	BranchName string
+}
+
+type unmarshalReviewersFn func(str string, client *github.Client) (map[string][]string, error)
+
+// CheckAndSetDefaults verifies configuration and sets defaults
+func (c *Config) CheckAndSetDefaults() error {
+	if c.Client == nil {
+		return trace.BadParameter("missing parameter Client")
+	}
+	if c.Reviewers == "" {
+		return trace.BadParameter("missing parameter Reviewers")
+	}
+	if c.EventPath == "" {
+		return trace.BadParameter("missing parameter EventPath")
+	}
+	if c.Token == "" {
+		return trace.BadParameter("missing parameter token")
+	}
+	if c.unmarshalRevs == nil {
+		c.unmarshalRevs = unmarshalReviewers
+	}
+	return nil
+}
+
+// New creates a new instance of Environment.
+func New(c Config) (*Environment, error) {
+
+	err := c.CheckAndSetDefaults()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	revs, err := c.unmarshalRevs(c.Reviewers, c.Client)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	pr, err := GetPullRequest(c.EventPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &Environment{
+		Client:           c.Client,
+		reviewers:        revs,
+		defaultReviewers: revs[""],
+		PullRequest:      pr,
+	}, nil
+}
+
+// unmarshalReviewers converts the passed in string representing json object into a map
+func unmarshalReviewers(str string, client *github.Client) (map[string][]string, error) {
+	var hasDefaultReviewers bool
+	if str == "" {
+		return nil, trace.NotFound("reviewers not found")
+	}
+	m := make(map[string][]string)
+
+	err := json.Unmarshal([]byte(str), &m)
+	if err != nil {
+		return nil, err
+	}
+	for author, requiredReviewers := range m {
+		for _, reviewer := range requiredReviewers {
+			_, err := userExists(reviewer, client)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+		if author == "" {
+			hasDefaultReviewers = true
+			continue
+		}
+		_, err := userExists(author, client)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+	if !hasDefaultReviewers {
+		return nil, trace.BadParameter("default reviewers are not set. set default reviewers with an empty string as a key")
+	}
+	return m, nil
+
+}
+
+// userExists checks if a user exists
+func userExists(userLogin string, client *github.Client) (*github.User, error) {
+	var targetUser *github.User
+	users, resp, err := client.Search.Users(context.TODO(), userLogin, &github.SearchOptions{})
```

Comment:
* Can you update these functions to take a context and pass the context through?
* Is this API paginated? It doesn't look like you're actually paginating the results here, the user may exist and you won't know.

Reply:
Done and ended up just using https://pkg.go.dev/github.com/google/go-github/v39/github#UsersService.Get which returns one user if they exist.